### DISCUSSION TITLE
FQDN and certificate generation

### DISCUSSION THREAD

**Discussion by Sahale** [Security-Onion-Solutions/securityonion#10992] [2.4] [answered]
Hello!

Installation of 2.4.4 (yes, it's already 2.4.5 but I missed it).

Errors during installation if use FQDN instead of hostname/IP while setting up access to SO.
```
Installing in non-interactive mode.{"log.level":"info","@timestamp":"2023-08-08T08:56:06.786Z","log.origin":{"file.name":"cmd/enroll_cmd.go","file.line":475},"message":"Starting enrollment to URL: https://blah.example.com:8220/","ecs.version":"1.6.0"}
                  Error: fail to enroll: fail to execute request to fleet-server: x509: certificate is valid for blah, not blah.example.com
```
Didn't find that it was fixed in 2.4.5.

**Comment by defensivedepth** [comment]
A few more technical details would be helpful:
- Install type (Standalone / EVAL / Manager / etc)
- ISO or Network Install
- Airgap?
- What is the technical context around needing FQDN?

**Reply by Sahale** [reply]
`Install type (Standalone / EVAL / Manager / etc)`
Distributed: Manager + Search

`ISO or Network Install`
ISO

`Airgap?`
No

`What is the technical context around needing FQDN?`
It's just more comfortable to use fqdn in local domain.

Thank you!

**Comment by Phillip-Samson** [comment]
I can attest to 
2.4 Standalone and Distributed
RC1 
Not Airgaped
The problem seems to not be fixed from my previous ticket
If you use just the hostname the cert will work

**Elastic Agent imbedded x509 for host url certificate only recognizes hostname not. full URL is there a way to only regenerate the new Elastic Agents #10504**

**Reply by jaypfe** [reply]
Distributed Install Type.
ISO.
Not airgap.

I need FQDN to enroll endpoints and ship logs from off network devices. I have a large amount of remote and traveling users who don't like to use the VPN.  Enrolling to a FQDN and having agents connect back to an FQDN allows me to do that. I have a public DNS pointing to my firewall IP (everything I have is NATed behind the edge router) then NATing back to my private IP of the SO. Then I just local resolve back to the box. 

Elastic agent logs being encrypted by default is very very nice. 

I see the installer updates with address changes, that's awesome! It does just seem the cert fails for any FQDN.

**Comment by defensivedepth** [comment]
Can you post a screenshot of `Fleet --> Settings` which shows the Fleet server hosts & Outputs?

**Comment by defensivedepth** [comment, marked as answer]
@Sahale @jaypfe @Bal33p 

I think I have figured it out. Appears that Elastic Fleet / Agent is not respecting the Cert CN when verifying the connection. Moving it to the SAN appears to have fixed it. Still testing, but if you would like to try it out, make the changes in this PR on your Manager:

https://github.com/Security-Onion-Solutions/securityonion/pull/11002

(That file is at `/opt/so/saltstack/default/salt/ssl/init.sls`). Then run the following:

`sudo salt-call state.highstate`

This should regenerate the certs & restart the containers.

**Reply by jaypfe** [reply]
That did it for me! Awesome job man! 
I did verify as well that the custom url's would still modify in the .exe and they do! 

I was able to enroll to fleet off-network and I am receiving logs with the FQDN. 

(I did have to ask the highstate to run 3 times because I kept getting a 'data couldn't compile' error but it eventually ran)

**Reply by defensivedepth** [reply]
Great to hear that @jaypfe !

Side note, if you have off-network users, I would suggest deploying a Fleet node, which is specifically designed for this use case, so that you don't have to expose the Manager to the public Internet.

**Reply by jaypfe** [reply]
Oh nice okay! I will start playing around with that. Thank you for the advice!

**Reply by Sahale** [reply]
Everything works now, thank you!

**Comment by Sahale** [comment]
One more thing about nginx certificates: old instructions to overwrite /etc/pki/managerssl.{crt,key} doesn't work, because so-nginx overwrites it while restart. How can I use my CA's certificates?
Moreover, it's generating only hostname in CN, so I temporarily modified `/opt/so/saltstack/local/salt/nginx/enabled.sls`:
```
managerssl_crt:
  x509.certificate_managed:
    - name: /etc/pki/managerssl.crt
    - ca_server: {{ ca_server }}
    - signing_policy: managerssl
    - private_key: /etc/pki/managerssl.key
    - CN: {{ GLOBALS.url_base }}
    - subjectAltName: DNS:{{ GLOBALS.hostname }}, IP:{{ GLOBALS.node_ip }}, DNS:{{ GLOBALS.url_base }}
```

**Reply by Sahale** [reply]
Oh, my fault, custom certificate for nginx is under "advanced" option in settings :)
